### DISCUSSION TITLE
Atom failing to launch for editing scripts

### DISCUSSION THREAD

**Discussion by cfritze64** [johnlindquist/kit#196] [Errors]
1.3.0-beta.6
setting up Kit on my second machine, a 2013 MacBook air running Big Sur
clean install of Kit and Atom
Atom has installed its Shell Commands
which atom in terminal gives /usr/local/bin/atom as expected
in the file ~/.kenv/.env is the line
KIT_EDITOR=/usr/local/bin/atom
the original version 
KIT_EDITOR=atom
didn't work either
from the terminal, I can run 
atom ~/.kenv/scripts/ascript    and atom launches with ascript.js ready for editing just fine.
if I run Kit from the command line and try to edit a script, I get the following error:

getScripts is not defined
ReferenceError: getScripts is not defined
    at Object.menu [as choices] (file:///Users/BikeFixe/.kit/cli/fns.js:11:9)
    at async global.kitPrompt (/Users/BikeFixe/.kit/preload/mac-terminal.cjs:498:19)
    at async global.arg (/Users/BikeFixe/.kit/preload/mac-terminal.cjs:544:15)
    at async file:///Users/BikeFixe/.kit/cli/edit.js?uuid=4fb0bd72-daf8-4fc9-b27c-37676ed7ab1c:3:12
{ line: '///Users/BikeFixe/.kit/cli/fns.js', col: '11' }

**Comment by johnlindquist** [comment]
Dang, sorry. I know why that `getScripts` error is occurring. I'll fix that tonight.

I'm looking into what I changed that disabled Atom. Will report back ASAP.

**Reply by johnlindquist** [reply]
Will you do me a favor to help confirm this fix:

Replace these two files in `~/.kit/preload` with the content of these gists:

[mac-app.cjs](https://gist.githubusercontent.com/johnlindquist/f6266d2035e8095eeae4c88295301fa2/raw/743dbce90a01a59048a6fbea70b235bda4307ca1/mac-app.cjs)

[mac-terminal.cjs](https://gist.githubusercontent.com/johnlindquist/db9da216f75de4b5c71e9a01079e0f29/raw/7b338aae547a3dfd2480d5af3a0b21da780847bf/mac-terminal.cjs)

**Reply by cfritze64** [reply]
replaced those files as directed. 
running kit from the terminal now lets me edits scripts fine, they launch in atom.
trying to edit the same scripts from the app interface fails to load atom (which I think you anticipated)
the kit app interface shows Opening Users/username.... when the edit function gets executed but it goes by too fast for me to capture or read it.

**Reply by johnlindquist** [reply]
Make sure .env is set to "atom", not the full path

**Reply by cfritze64** [reply]
I changed .env so just atom is assigned to KIT_EDITOR
atom ~/.kenv/scripts/scriptname    opens script in atom
if I try to edit the script from the kit cli I get:

{
  command: 'atom "/Users/BikeFixe/.kenv/scripts/testee.js" "/Users/BikeFixe/.kenv"'
}
mkdir: /.atom: Read-only file system
[String: ''] {
  stdout: '',
  stderr: 'mkdir: /.atom: Read-only file system\n',
  code: 0,
  cat: [Function: bound ],
  exec: [Function: bound ],
  grep: [Function: bound ],
  head: [Function: bound ],
  sed: [Function: bound ],
  sort: [Function: bound ],
  tail: [Function: bound ],
  to: [Function: bound ],
  toEnd: [Function: bound ],
  uniq: [Function: bound ]
}
> Opening /Users/BikeFixe/.kenv/scripts/testee.js with atom

**Reply by johnlindquist** [reply]
Does it open with the App?

**Reply by cfritze64** [reply]
No

**Reply by cfritze64** [reply]
in summary: replacing the two .cjs files restored function to the terminal interface while my .env file had the full path for atom.

using just "atom" for the editor path in .env introduced this new error in the terminal interface.

In both cases, the app interface failed to open the scripts for editing

**Reply by johnlindquist** [reply]
Interesting.

That error is coming from here in the Atom (not Kit) launch script:

https://github.com/atom/atom/blob/master/atom.sh#L91

I must need to pass a $HOME env var, because the error looks like it's trying to create a directory at the root of your system.

**Reply by cfritze64** [reply]
feel free to move this down your priority list. I can get at the scripts in other ways for editing...

**Reply by johnlindquist** [reply]
No way, you're my priority. You've supported me since the beginning :)

**Reply by cfritze64** [reply]
OK :-)
I kept banging on it. New behavior from the terminal. when I try to edit script from the kit cli interface, I get that error above in the terminal window, then after a brief hesitation, the script opens for editing in atom

**Reply by johnlindquist** [reply]
That entire atom.sh file is to build up this one line: https://github.com/atom/atom/blob/master/atom.sh#L150

```bash
open -a "$ATOM_PATH/$ATOM_APP_NAME" -n --args --executed-from="$(pwd)" --pid=$$ --path-environment="$PATH" "$@"
```
I may just hardcode that with the appropriate variables since we know we're on mac, which files we're opening, etc.

**Reply by cfritze64** [reply]
>>That entire atom.sh file is to build up this one line:[...]

Woof! Or can you use your own open() or exec() functions to issue a simple `atom ${path}${filename}` command or some such?

**Reply by johnlindquist** [reply]
Calling exec(atom) calls that script. That's the issue. Providing the correct env vars when calling exec

**Reply by johnlindquist** [reply]
Meaning, that's the script at /usr/local/bin/atom

**Reply by cfritze64** [reply]
aha! got it

**Reply by johnlindquist** [reply]
Can you run "Check for update" from the help menu? I think I got it. I'm passing the $HOME var properly now. (works on my machine 😅)

**Reply by cfritze64** [reply]
Thanks much. I’m In a car all day. Will test tonight

**Reply by cfritze64** [reply]
just tested with beta7. from the app interface, no joy getting atom to launch on script edit. 
from the terminal CLI kit, edit script opens the script in atom, but/and terminal also gives this response before the open issues:

[String: ''] {
  stdout: '',
  stderr: '',
  code: 0,
  cat: [Function: bound ],
  exec: [Function: bound ],
  grep: [Function: bound ],
  head: [Function: bound ],
  sed: [Function: bound ],
  sort: [Function: bound ],
  tail: [Function: bound ],
  to: [Function: bound ],
  toEnd: [Function: bound ],
  uniq: [Function: bound ]
}
> Opening /Users/BikeFixe/.kenv/scripts/testee.js with atom

**Reply by cfritze64** [reply]
more weirdness, when I try to select scripts in the edit command of the kit cli interface, kit often reports "script not found"

**Reply by cfritze64** [reply]
OK. is there some sort of permissions issue? here's my prompt in terminal and the result of 
ls -l

Christians-MacBook-Air:scripts BikeFixe$ ls -l
total 208
-rw-r--r--  1 BikeFixe  staff  1227 Apr 27 07:52 app-launcher.js
-rw-r--r--@ 1 BikeFixe  staff  1646 Apr 27 07:52 bcco-menu.js
-rw-r--r--  1 BikeFixe  staff   160 Apr 27 07:52 browse-scriptkit.js
-rw-r--r--  1 BikeFixe  staff   543 Apr 27 07:52 center-app.js
-rw-r--r--  1 BikeFixe  staff   800 Apr 26 14:38 clipboard-history.js
-rw-r--r--  1 BikeFixe  staff   338 Apr 27 07:52 curl-simplelists.js
-rw-r--r--  1 BikeFixe  staff   342 Apr 26 14:38 dad-joke.js
-rw-r--r--  1 BikeFixe  staff   221 Apr 27 07:52 edit-main.js
-rw-r--r--  1 BikeFixe  staff   478 Apr 27 07:52 google-it.js
-rw-r--r--  1 BikeFixe  staff    50 Apr 27 07:52 hotkey-demo.js
-rw-r--r--  1 BikeFixe  staff   523 Apr 27 07:52 left-app.js
-rw-r--r--  1 BikeFixe  staff   233 Apr 27 07:52 phone.js
-rw-r--r--  1 BikeFixe  staff  6398 Apr 27 07:52 piper-script.js
-rw-r--r--  1 BikeFixe  staff  1688 Apr 27 07:52 reddit.js
-rw-r--r--  1 BikeFixe  staff   500 Apr 27 07:52 right-app.js
-rw-r--r--  1 BikeFixe  staff   244 Apr 26 14:38 run-tutorial.js
-rw-r--r--  1 BikeFixe  staff   845 Apr 27 07:52 sandbox.js
-rw-r--r--  1 BikeFixe  staff  1520 Apr 27 07:52 set-zoom-windows-rehearsal.js
-rw-r--r--  1 BikeFixe  staff  1520 Apr 27 07:52 set-zoom-windows.js
-rw-r--r--  1 BikeFixe  staff  1412 Apr 27 07:52 test-script.js
-rw-r--r--@ 1 BikeFixe  staff    12 Apr 26 16:10 testee.js
-rw-r--r--  1 BikeFixe  staff   755 Apr 26 14:38 todos.js
-rw-r--r--  1 BikeFixe  staff   579 Apr 27 07:52 truve.js
-rw-r--r--  1 BikeFixe  staff  1212 Apr 26 14:38 word-api.js
-rw-r--r--  1 BikeFixe  staff   950 Apr 27 07:52 zoom-quick-links.js

the scripts with the trailing @ symbol in their permissions listing run and can be edited from the kit cli interface. Those without the trailing @ give script not found errors when I try to run or edit them in the kit cli interface.

If you want me to entirely wipe my .profile file, the .kenv and .kit directories in my home director, and atom and start all over, I can do that...

**Reply by johnlindquist** [reply]
Let me dig into it first. I was doing some menu caching to help speed up the initial App launch. This may have inadvertently impacted the cli menu.

**Reply by johnlindquist** [reply]
I'm pretty sure I got it this time if you "Help->Check for update". 🤞

**Reply by cfritze64** [reply]
Super. Two improvements and one remaining quirk.

if I run kit from the terminal, editing a script works great - opens atom right up to the selected script.
Same thing in the kit app - editing opens atom to the selected script.

The quirk: something still strange with "run a script" in the terminal app. I run kit in the terminal and type "run" and return to select the "run: Run a script" function. This leaves kit waiting for input on which script to run. If I type "dad" and return, I see that dad-joke is selected in response to my typing, but when I hit that return key, I get a message in the terminal "❯ Script Dad Joke not found. Please select a different script:". Same happens if I try to select dad-joke using arrow keys instead of typing "dad", or with other scripts.

**Reply by johnlindquist** [reply]
Oh, cool, glad to hear that Atom is finally working, phew.

That "Script not found" (previously from edit, now on run) is caused by the same validator failing. I wrote the validator and only tested it in the App, but the terminal behaves a tiny differently, so I need to adjust the validator to work with both. I'll get this fixed tomorrow.


FWIW, you should be able to run `kit run dad-joke` and skip the menu completely if you know the exact name of the script.
